torchrl/objectives/value/vtrace.py

```diff
@@ -4,55 +4,334 @@
+    deltas = clipped_rho * (rewards + gamma * next_vals - vals)
+    return deltas, clipped_rho
 
 
-def _vtrace(
-    rewards: torch.Tensor,
-    vals: torch.Tensor,
-    log_pi: torch.Tensor,
-    log_mu: torch.Tensor,
-    gamma: Union[torch.Tensor, float],
-    rho_bar: Union[float, torch.Tensor] = 1.0,
-    c_bar: Union[float, torch.Tensor] = 1.0,
+# @_transpose_time # TODO: is this needed?
+def vtrace_correction(
+        gamma: float,
+        log_pi: torch.Tensor,
+        log_mu: torch.Tensor,
+        state_value: torch.Tensor,
+        next_state_value: torch.Tensor,
+        reward: torch.Tensor,
+        done: torch.Tensor,
+        rho_thresh: Union[float, torch.Tensor] = 1.0,
+        c_thresh: Union[float, torch.Tensor] = 1.0,
+        time_dim: int = -2,
 ) -> Tuple[torch.Tensor, torch.Tensor]:
-    T = vals.shape[1]
-    if not isinstance(gamma, torch.Tensor):
-        gamma = torch.full_like(vals, gamma)
-
-    dv, rho = _dv_val(rewards, vals, gamma, rho_bar, log_pi, log_mu)
-    c = _c_val(log_pi, log_mu, c_bar)
-
-    v_out = []
-    v_out.append(vals[:, -1] + dv[:, -1])
-    for t in range(T - 2, -1, -1):
-        _v_out = (
-            vals[:, t] + dv[:, t] + gamma[:, t] * c[:, t] * (v_out[-1] - vals[:, t + 1])
+    """Computes V-Trace off-policy actor critic targets.
+
+    Refer to "IMPALA: Scalable Distributed Deep-RL with Importance Weighted  Actor-Learner Architectures"
+    https://arxiv.org/abs/1802.01561 for more context.
+
+    Args:
+        gamma (scalar): exponential mean discount.
+        log_pi (Tensor): log probability of taking actions in the environment.
+        log_mu (Tensor): log probability of taking actions in the environment.
+        state_value (Tensor): value function result with old_state input.
+        next_state_value (Tensor): value function result with new_state input.
+        reward (Tensor): reward of taking actions in the environment.
+        done (Tensor): boolean flag for end of episode.
+        rho_thresh (Union[float, Tensor]): clipping parameter for importance weights.
+        c_thresh (Union[float, Tensor]): clipping parameter for importance weights.
+        time_dim (int): dimension where the time is unrolled. Defaults to -2.
+
+    All tensors (values, reward and done) must have shape
+    ``[*Batch x TimeSteps x *F]``, with ``*F`` feature dimensions.
+
+    """
+
+    if not (next_state_value.shape == state_value.shape == reward.shape == done.shape):
+        raise RuntimeError(SHAPE_ERR)
+
+    dtype = next_state_value.dtype
+    device = state_value.device
+
+    deltas, clipped_rho = _dv_val(reward, state_value, next_state_value, gamma, rho_thresh, log_pi, log_mu)
+    clipped_c = torch.min(torch.tensor(c_thresh).to(device), clipped_rho)
+
+    ############################################################
+    # MAKE THIS PART WORK; THEN WE CAN TRY TO MAKE IT FASTER
+
+    not_done = (~done).int()
+    *batch_size, time_steps, lastdim = not_done.shape
```

Comment:
if you don't implement @_transpose_time this will break when people pass `time_dim`